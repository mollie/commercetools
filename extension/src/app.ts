import express, { Application } from 'express';
import handler from '../expressHandler';
import morganMiddleware from './logger/morganMiddleware';

const app: Application = express();

app.use(morganMiddleware); // Logs from http level

app.get('/health', (req, res) => {
  res.send('Ok');
});

app.use(express.json());

app.use('/', handler);

export default app;
