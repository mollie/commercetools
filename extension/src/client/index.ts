import createMollieClient, { MollieClient } from '@mollie/api-client';
import { ClientBuilder, Client, AuthMiddlewareOptions, HttpMiddlewareOptions } from '@commercetools/sdk-client-v2';
import config from '../../config/config';
import { version } from '../../package.json';

export function initialiseMollieClient(): MollieClient {
  const mollieApiKey = config.mollie.apiKey;
  const mollieUserAgentString = `MollieCommercetools-extension/${version}`;
  const mollieClient = createMollieClient({ apiKey: mollieApiKey, versionStrings: mollieUserAgentString });
  return mollieClient;
}

export function initialiseCommercetoolsClient(): Client {
  const {
    commercetools: { projectKey, clientId, clientSecret, host, authUrl, scopes },
  } = config;

  const authMiddlewareOptions: AuthMiddlewareOptions = {
    host: authUrl,
    projectKey,
    credentials: {
      clientId,
      clientSecret,
    },
    scopes,
    fetch,
  };

  const httpMiddlewareOptions: HttpMiddlewareOptions = {
    host,
    fetch,
  };

  const client: Client = new ClientBuilder()
    .withClientCredentialsFlow(authMiddlewareOptions)
    .withHttpMiddleware(httpMiddlewareOptions)
    .withUserAgentMiddleware()
    .build();

  return client;
}
