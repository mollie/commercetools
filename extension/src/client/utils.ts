import createMollieClient, { MollieClient } from '@mollie/api-client';
import config from '../../config/config';
import { version } from '../../package.json';

export function initialiseMollieClient(): MollieClient {
  const mollieApiKey = config.mollie.apiKey;
  const mollieUserAgentString = `MollieCommercetools-extension/${version}`;
  const mollieClient = createMollieClient({ apiKey: mollieApiKey, versionStrings: mollieUserAgentString });
  return mollieClient;
}
