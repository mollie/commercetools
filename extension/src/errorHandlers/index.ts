import { formatMollieErrorResponse } from './formatMollieErrorResponse';
import { formatExtensionErrorResponse } from './formatExtensionErrorResponse';

const errorHandler = {
  mollie: formatMollieErrorResponse,
  extension: formatExtensionErrorResponse,
};

export default errorHandler;
