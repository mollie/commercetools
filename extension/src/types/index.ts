// This is not exhaustive
// If you use another commercetools error response code, add it to this enum
export enum CTEnumErrors {
  General = 'General',
  InvalidInput = 'InvalidInput',
  InvalidOperation = 'InvalidOperation',
  Unauthorized = 'Unauthorized',
  SyntaxError = 'SyntaxError',
  SemanticError = 'SemanticError',
  ObjectNotFound = 'ObjectNotFound',
}

export type CTUpdatesRequestedResponse = {
  status: number;
  actions?: Action[];
  errors?: CTError[];
};

export type Action = {
  action: string;
  type?: {
    key: string;
  };
  fields?: {
    actionType: string;
    request?: string;
    response?: string;
    createdAt?: string;
  };
  name?: string;
  value?: string;
  key?: string;
  transaction?: CTTransaction;
  transactionId?: string;
  interactionId?: string;
  state?: CTTransactionState;
};

export type CTError = {
  code: CTEnumErrors;
  message: string;
  extensionExtraInfo?: Object;
};

export type CTMoney = {
  type?: 'centPrecision';
  currencyCode: string;
  centAmount: number;
  fractionDigits?: number;
};

export type CTPayment = {
  id: string;
  amountPlanned: CTMoney;
  transactions?: CTTransaction[];
  key?: string;
  custom?: {
    fields?: {
      paymentMethodsRequest?: string;
      paymentMethodsResponse?: string;
      createPayment?: string;
      createCapture?: string;
      cancelPayment?: string;
      createRefund?: string;
    };
  };
  interfaceInteractions?: CTInterfaceInteraction[];
};

export type CTInterfaceInteraction = {
  id: string;
  actionType: string;
  createdAt?: Date;
  request?: string;
  response?: string;
};

export type CTTransaction = {
  id?: string;
  timestamp?: string;
  type: CTTransactionType;
  amount: CTMoney;
  interactionId?: string;
  state?: 'Initial' | 'Pending' | 'Success' | 'Failure';
};

export enum ControllerAction {
  GetPaymentMethods = 'getPaymentMethods',
  CreateOrder = 'createOrder',
  CreateOrderPayment = 'createOrderPayment',
  CreateOrderRefund = 'createOrderRefund',
  CreateCustomRefund = 'createCustomRefund',
  CreateShipment = 'createShipment',
  UpdateShipment = 'updateShipment',
  CancelOrder = 'cancelOrder',
  NoAction = 'noAction',
  Error = 'error',
}

export enum CTTransactionType {
  Authorization = 'Authorization',
  CancelAuthorization = 'CancelAuthorization',
  Charge = 'Charge',
  Refund = 'Refund',
  Chargeback = 'Chargeback',
}

export enum CTTransactionState {
  Initial = 'Initial',
  Pending = 'Pending',
  Success = 'Success',
  Failure = 'Failure',
}
