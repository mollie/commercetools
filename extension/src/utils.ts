import { MethodsListParams } from '@mollie/api-client';
/**
 * Generates an ISO string date
 * @returns {String} Returns the current date converted to ISO.
 */
export function createDateNowString() {
  return new Date().toISOString();
}

export function amountMapper(amountPlanned: any): string {
  const { centAmount, fractionDigits } = amountPlanned;
  const divider = Math.pow(10, fractionDigits || 2);
  const mollieAmount = (centAmount / divider).toFixed(2);
  return mollieAmount
}

export function methodListMapper(ctObj: any): MethodsListParams {
  // Generally this shouldn't be needed, but a safety anyway.. eventually could return error here
  if (!ctObj.amountPlanned) {
    return {};
  }
  const mollieAmount = amountMapper(ctObj.amountPlanned);

  const mObject: MethodsListParams = {
    amount: {
      value: mollieAmount,
      currency: ctObj.amountPlanned.currencyCode,
    },
    // Resource is hardcoded, for the time being we only support Orders API
    resource: 'orders',
  };

  if (ctObj.custom?.fields?.paymentMethodsRequest) {
    const parsedMethodsRequest = JSON.parse(ctObj.custom?.fields?.paymentMethodsRequest)
    const { locale, billingCountry, includeWallets, orderLineCategories, issuers, pricing, sequenceType } = parsedMethodsRequest;
    const include = issuers || pricing ? `${issuers ? 'issuers,' : ''}${pricing ? 'pricing' : ''}` : undefined;

    Object.assign(
      mObject,
      locale && { locale: locale },
      include && { include: include },
      includeWallets && { includeWallets: includeWallets },
      billingCountry && { billingCountry: billingCountry },
      sequenceType && { sequenceType: sequenceType },
      orderLineCategories && { orderLineCategories: orderLineCategories },
    );
  }

  return mObject;
}
