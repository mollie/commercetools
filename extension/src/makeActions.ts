import { time } from 'console';
import { v4 as uuid } from 'uuid';
import { Action, ControllerAction, CTTransactionState } from './types';
import { createDateNowString } from './utils';

/**
 *
 * @param customFieldName
 * @param customFieldValue
 * If the customFieldValue is an API response, JSON Stringify it before passing it
 */
const setCustomField = (customFieldName: string, customFieldValue: string) => {
  return {
    action: 'setCustomField',
    name: customFieldName,
    value: customFieldValue,
  };
};

type createInterfaceInteractionParams = {
  actionType: ControllerAction;
  requestValue: string;
  responseValue: string;
  id?: string;
  timestamp?: string;
};
/**
 * @param parameters type createInterfaceInteractionParams, which contains
 * actionType ControllerAction
 * requestValue string
 * responseValue string
 * id string string, optional
 * timestamp string, optional
 * If the responseValue is an API response, JSON Stringify it before passing it
 */
const addInterfaceInteraction = (params: createInterfaceInteractionParams) => {
  const { actionType, requestValue, responseValue, id, timestamp } = params;
  const interfaceInteractionId = id ? id : uuid();
  const interfaceInteractionTimestamp = timestamp ? timestamp : createDateNowString();
  return {
    action: 'addInterfaceInteraction',
    type: {
      key: 'ct-mollie-integration-interface-interaction-type',
    },
    fields: {
      id: interfaceInteractionId,
      actionType,
      createdAt: interfaceInteractionTimestamp,
      request: requestValue,
      response: responseValue,
    },
  };
};

/**
 *
 * @param key will be the mollie order id
 */
const setKey = (key: string) => {
  return {
    action: 'setKey',
    key,
  };
};

/**
 *
 * @param interfaceText will be the mollie order status
 */
const setStatusInterfaceText = (interfaceText: string) => {
  return {
    action: 'setMethodInfoName',
    interfaceText,
  };
};

/**
 *
 * @param id transaction to be updated
 * @param newState CTTransactionState
 */
const changeTransactionState = (id: string, newState: CTTransactionState) => {
  return {
    action: 'changeTransactionState',
    transactionId: id,
    state: newState,
  };
};

/**
 *
 * @param id transaction to be updated
 * @param interactionId either the mollie payment id, or the corresponding interfaceInteraction's id
 */
const changeTransactionInteractionId = (id: string, interactionId: string) => {
  return {
    action: 'changeTransactionInteractionId',
    transactionId: id,
    interactionId,
  };
};

/**
 *
 * @param id transaction to be updated
 * @param timestamp CT DateTime is a JSON string representation of UTC date & time in ISO 8601 format (YYYY-MM-DDThh:mm:ss.sssZ)
 * for example: "2018-10-12T14:00:00.000Z"
 */
const changeTransactionTimestamp = (id: string, timestamp: string) => {
  return {
    action: 'changeTransactionTimestamp',
    transactionId: id,
    timestamp,
  };
};

export const makeActions = {
  setCustomField,
  addInterfaceInteraction,
  setKey,
  setStatusInterfaceText,
  changeTransactionState,
  changeTransactionInteractionId,
  changeTransactionTimestamp,
};
