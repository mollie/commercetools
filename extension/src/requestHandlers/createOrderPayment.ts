import { MollieClient, OrderPaymentCreateParams, Payment } from '@mollie/api-client';
import { formatMollieErrorResponse } from '../errorHandlers/formatMollieErrorResponse';
import { Action, CTTransactionType, CTUpdatesRequestedResponse } from '../types';
import { createDateNowString } from '../utils';
import Logger from '../logger/logger';

export function getOrdersPaymentsParams(ctObj: any): Promise<OrderPaymentCreateParams> {
  try {
    const parsedCreateOrderPaymentRequest = JSON.parse(ctObj?.custom?.fields?.createOrderPaymentRequest);
    const orderPaymentCreateParams = {
      orderId: ctObj?.key,
    };
    const { method, customerId, mandateId } = parsedCreateOrderPaymentRequest;
    if (method) Object.assign(orderPaymentCreateParams, { method });
    if (customerId) Object.assign(orderPaymentCreateParams, { customerId });
    if (mandateId) Object.assign(orderPaymentCreateParams, { mandateId });
    return Promise.resolve(orderPaymentCreateParams);
  } catch (error) {
    Logger.error({ error });
    return Promise.reject({ status: 400, title: 'Could not make parameters needed to create Mollie order payment.', field: 'createOrderPaymentRequest' });
  }
}

export function createCtActions(orderPaymentResponse: Payment, ctObj: any): Action[] {
  const stringifiedOrderPaymentResponse = JSON.stringify(orderPaymentResponse);
  const result: Action[] = [
    {
      action: 'addInterfaceInteraction',
      type: {
        key: 'ct-mollie-integration-interface-interaction-type',
      },
      fields: {
        actionType: 'createOrderPayment',
        createdAt: createDateNowString(),
        request: ctObj?.custom?.fields?.createOrderPaymentRequest,
        response: stringifiedOrderPaymentResponse,
      },
    },
    {
      action: 'setCustomField',
      name: 'createOrderPaymentResponse',
      value: stringifiedOrderPaymentResponse,
    },
    {
      action: 'addTransaction',
      transaction: {
        timestamp: orderPaymentResponse.createdAt,
        type: CTTransactionType.Charge,
        amount: ctObj.amountPlanned,
        interactionId: orderPaymentResponse.id,
      },
    },
  ];
  return result;
}

export default async function createOrderPayment(ctObj: any, mollieClient: MollieClient, getOrdersPaymentsParams: Function, createCtActions: Function): Promise<CTUpdatesRequestedResponse> {
  try {
    const ordersPaymentsParams = await getOrdersPaymentsParams(ctObj);
    const mollieOrderPaymentRes = await mollieClient.orders_payments.create(ordersPaymentsParams);
    const ctActions = createCtActions(mollieOrderPaymentRes, ctObj);
    return {
      actions: ctActions,
      status: 201,
    };
  } catch (error: any) {
    Logger.error({ error });
    const errorResponse = formatMollieErrorResponse(error);
    return errorResponse;
  }
}
