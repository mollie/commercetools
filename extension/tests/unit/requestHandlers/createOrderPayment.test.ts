import { Payment } from '@mollie/api-client';
import { v4 as uuid } from 'uuid';
import { mocked } from 'ts-jest/utils';
import createOrderPayment, { getOrdersPaymentsParams, createCtActions } from '../../../src/requestHandlers/createOrderPayment';
import { Action, CTPayment } from '../../../src/types';
import { makeActions } from '../../../src/makeActions';
import Logger from '../../../src/logger/logger';

jest.mock('../../../src/makeActions');
jest.mock('uuid');

describe('getOrdersPaymentsParams', () => {
  it('Should create required params for mollie createOrderPayment call', async () => {
    const mockCtPayment = {
      key: 'ord_3uwvfd',
      paymentMethodInfo: {
        paymentInterface: 'Mollie',
        method: 'ideal',
      },
    } as CTPayment;
    const expectedCreateOrderPaymentParams = {
      orderId: 'ord_3uwvfd',
      method: 'ideal',
    };
    expect(getOrdersPaymentsParams(mockCtPayment)).toEqual(expectedCreateOrderPaymentParams);
  });
});

describe('createCtActions', () => {
  it('Should create correct ct actions from request and mollies payment response', async () => {
    const mockCtPayment = {
      key: 'ord_3uwvfd',
      paymentMethodInfo: {
        paymentInterface: 'Mollie',
        method: 'ideal',
      },
      transactions: [
        {
          id: 'b8243016-fc80-4af8-a273-47801f72ac31',
          timestamp: '2022-01-10T07:14:27.000Z',
          type: 'Charge',
          amount: {
            type: 'centPrecision',
            currencyCode: 'EUR',
            centAmount: 1104,
            fractionDigits: 2,
          },
          interactionId: 'tr_HfGsMew6vQ',
          state: 'Initial',
        },
      ],
    } as CTPayment;
    const mockOrderPaymentRes = {
      resource: 'payment',
      id: 'tr_mAmMrPGnxe',
      createdAt: '2021-10-20T15:35:13+00:00',
      amount: { value: '4.20', currency: 'EUR' },
      description: 'Order 1001',
      method: null,
      status: 'open',
      profileId: 'pfl_VtWA783A63',
      orderId: 'ord_3uwvfd',
    } as any as Payment;
    const ctActions = await createCtActions(mockOrderPaymentRes, mockCtPayment);
    expect(ctActions).toHaveLength(3);
    expect(makeActions.addInterfaceInteraction).toHaveBeenCalledTimes(1);
    expect(makeActions.changeTransactionInteractionId).toHaveBeenCalledTimes(1);
    expect(makeActions.changeTransactionInteractionId).toHaveBeenCalledWith(mockCtPayment.transactions![0].id, mockOrderPaymentRes.id);
    expect(makeActions.changeTransactionTimestamp).toHaveBeenCalledTimes(1);
    expect(makeActions.changeTransactionTimestamp).toHaveBeenCalledWith(mockCtPayment.transactions![0]?.id, mockOrderPaymentRes.createdAt);
  });
  it('Should return an error if no initial transaction is found', async () => {
    const mockCtPayment = {
      key: 'ord_3uwvfd',
      paymentMethodInfo: {
        paymentInterface: 'Mollie',
        method: 'ideal',
      },
      transactions: [
        {
          id: 'b8243016-fc80-4af8-a273-47801f72ac31',
          timestamp: '2022-01-10T07:14:27.000Z',
          type: 'Charge',
          amount: {
            type: 'centPrecision',
            currencyCode: 'EUR',
            centAmount: 1104,
            fractionDigits: 2,
          },
          interactionId: 'tr_HfGsMew6vQ',
          state: 'Pending',
        },
      ],
    } as CTPayment;
    const expectedError = { status: 400, title: 'Cannot find original transaction', field: 'Payment.transactions' };
    await expect(createCtActions({} as any as Payment, mockCtPayment)).rejects.toMatchObject(expectedError);
  });
  it('Should return an error when generating actions fails', async () => {
    mocked(uuid).mockImplementationOnce(() => {
      throw new Error('Test error');
    });
    const mockCtPayment = {
      key: 'ord_3uwvfd',
      paymentMethodInfo: {
        paymentInterface: 'Mollie',
        method: 'ideal',
      },
      transactions: [
        {
          id: 'b8243016-fc80-4af8-a273-47801f72ac31',
          timestamp: '2022-01-10T07:14:27.000Z',
          type: 'Charge',
          amount: {
            type: 'centPrecision',
            currencyCode: 'EUR',
            centAmount: 1104,
            fractionDigits: 2,
          },
          interactionId: 'tr_HfGsMew6vQ',
          state: 'Initial',
        },
      ],
    } as CTPayment;
    await expect(createCtActions({} as any as Payment, mockCtPayment)).rejects.toEqual(Error('Test error'));
  });
});
describe('createOrderPayment', () => {
  const mockLogError = jest.fn();
  beforeEach(() => {
    Logger.error = mockLogError;
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('Should prepare params, call mollie, handle response and return actions', async () => {
    const mockOrdersPaymentsParams = { orderId: 'ord_3uwvfd', method: 'ideal' };
    const mockCtPayment = {
      key: 'ord_3uwvfd',
      paymentMethodInfo: {
        paymentInterface: 'Mollie',
        method: 'ideal',
      },
    } as CTPayment;
    const mockOrderPaymentResponse: any = {
      resource: 'payment',
      id: 'tr_mAmMrPGnxe',
      createdAt: '2021-10-20T15:35:13+00:00',
      amount: { value: '4.20', currency: 'EUR' },
      description: 'Test Order',
      method: null,
      status: 'open',
      orderId: 'ord_3uwvfd',
    };
    const mockCtActions: Action[] = [];
    const getOrdersPaymentsParams = jest.fn().mockReturnValueOnce(mockOrdersPaymentsParams);
    const createCtActions = jest.fn().mockResolvedValueOnce(mockCtActions);
    const mollieClient = { orders_payments: { create: jest.fn().mockResolvedValueOnce(mockOrderPaymentResponse) } } as any;

    const createOrderPaymentRes = await createOrderPayment(mockCtPayment, mollieClient, getOrdersPaymentsParams, createCtActions);
    expect(getOrdersPaymentsParams).toBeCalledWith(mockCtPayment);
    expect(mollieClient.orders_payments.create).toHaveBeenCalledWith(mockOrdersPaymentsParams);
    expect(createCtActions).toBeCalledWith(mockOrderPaymentResponse, mockCtPayment);
    expect(createOrderPaymentRes.status).toBe(201);
  });
  it('Should return commercetools formated error if one of the functions fails', async () => {
    const mockOrdersPaymentsParams = { orderId: 'ord_3uwvfd', method: 'ideal' };
    const mockCtPayment = {
      key: 'ord_3uwvfd',
      paymentMethodInfo: {
        paymentInterface: 'Mollie',
        method: 'ideal',
      },
    } as CTPayment;
    const mockError = { statusCode: 404, title: 'Not found', message: 'No order exists with token ord_3uwvfd', field: undefined };
    const getOrdersPaymentsParams = jest.fn().mockReturnValueOnce(mockOrdersPaymentsParams);
    const createCtActions = jest.fn();
    const mollieClient = { orders_payments: { create: jest.fn().mockRejectedValueOnce(mockError) } } as any;

    const createOrderPaymentRes = await createOrderPayment(mockCtPayment, mollieClient, getOrdersPaymentsParams, createCtActions);
    expect(createOrderPaymentRes.status).toBe(400);
    expect(createOrderPaymentRes.errors).toHaveLength(1);
    expect(createOrderPaymentRes).toMatchSnapshot();
  });
});
