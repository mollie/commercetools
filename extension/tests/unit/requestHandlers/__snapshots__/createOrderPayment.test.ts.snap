// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createOrderPayment Should return commercetools formated error if one of the functions fails 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "ObjectNotFound",
      "extensionExtraInfo": Object {
        "originalStatusCode": 404,
        "title": "Not found",
      },
      "message": "No order exists with token ord_3uwvfd",
    },
  ],
  "status": 400,
}
`;
